iCount/analysis/rnamaps.py

```diff
@@ -3,690 +3,233 @@
-          of first-stop-s. THe major question is: "Do all reads for certain
-          xlink event indicate the same behaviour?"
-
+Perform RNA-maps analysis.
 """
+import csv
 import logging
+import os
 
-import pybedtools
-import matplotlib
-matplotlib.use('Agg')
-import matplotlib.pyplot as plt  # pylint: disable=wrong-import-position
+from pybedtools import BedTool, create_interval_from_list
 
 import iCount  # pylint: disable=wrong-import-position
-from iCount.files import _f2s  # pylint: disable=wrong-import-position
 
 LOGGER = logging.getLogger(__name__)
 
-EXON_TYPES = ['CDS', 'ncRNA', 'UTR3', 'UTR5']
-RNA_WINDOW_SIZE = 2000  # TODO: rethink how this constant would affect lengths in RNAmap generation
-
-
-def _add_entry(start_type, stop_type, distance, score, strand, data, metrics, explict=False):
-    """Add RNA-map entry in ``data``."""
-    if strand == '+':
-        rna_map_type = start_type + '-' + stop_type
-    else:
-        rna_map_type = stop_type + '-' + start_type
-        distance = -distance
-
-    data.setdefault(rna_map_type, {}).setdefault(distance, [0, 0])[0] += score
-    if explict:
-        data.setdefault(rna_map_type, {}).setdefault(distance, [0, 0])[1] += score
-
-    # Increment also mRNA / pre-mRNa counters:
-    if 'intron' in rna_map_type or 'intergenic' in rna_map_type:
-        metrics.origin_premrna += score
-    elif start_type in EXON_TYPES and stop_type in EXON_TYPES:
-        metrics.origin_mrna += score
-    else:
-        metrics.origin_ambiguous += score
-
-
-def _process_read_group(xlink, chrom, strand, read, data, segmentation, metrics,
-                        implicit_handling='closest'):
-    """
-    Process each read group.
-
-    Each group is represented by read. It has a fixed xlink, start, second-start
-    and end position. This function finds all landmarks that this read is
-    overlapping. It also determines for each of them:
+MAP_TYPES = {
+    'exon-intron': {
+        'upstream-type': ['CDS'],
+        'upstream-size-limit': 50,
+        'upstream-plot-width': 50,
+        'downstream-type': ['intron'],
+        'downstream-size-limit': 150,
+        'downstream-plot-width': 100,
+    },
+    'intron-exon': {
+        'upstream-type': ['intron'],
+        'upstream-size-limit': 150,
+        'upstream-plot-width': 100,
+        'downstream-type': ['CDS'],
+        'downstream-size-limit': 50,
+        'downstream-plot-width': 50,
+    },
+    'gene-start': {
+        'upstream-type': ['intergenic'],
+        'upstream-plot-width': 200,
+        'downstream-type': ['UTR5', 'CDS'],
+        'downstream-plot-width': 50,
+    },
+    'gene-end': {
+        'upstream-type': ['CDS', 'UTR3'],
+        'upstream-plot-width': 200,
+        'downstream-type': ['intergenic'],
+        'downstream-plot-width': 200,
+    },
+    'translation-start': {
+        'upstream-type': ['UTR5'],
+        'upstream-plot-width': 50,
+        'downstream-type': ['intron', 'CDS'],
```

Comment:
@JureZmrzlikar correct to: 'downstream-type': ['CDS']

Reply:
Are you sure? It believe the most common sequence of types at transcription start is exactly `UTR5`-`intron`-`CDS` This would leave out the majority of cases...

Reply:
Here we wish to analyse the boundary between UTR5 and CDR, for this they need to be present in the same exon without an intervening intron. Indeed, this will leave out many cases, because either UTR5 or CDS are often too short when in the same exon. For this, we could set the size-limit to less than 50 (maybe 20), to allow more cases.

Reply:
Just checked: there are 12196 of all such cases, if we set size limit to 50/50 there are only 6641 left. If we set limits to 20/20 there are 9754 of them.

Reply:
All this is for human, GENCODE annotation.